udata/features/territories/views.py

```diff
@@ -109,19 +109,39 @@ def render_territory(territory):
     if not current_app.config.get('ACTIVATE_TERRITORIES'):
         return abort(404)
 
-    DATASETS = TERRITORY_DATASETS[territory.level_name]
+    # Check that this territory is still valid (redirect old French
+    # regions for instance).
+    if territory.validity.get('end'):
+        valid_territory = GeoZone.objects.valid().get(
+            level=territory.level,
+            ancestors__contains=territory.id)
+        if valid_territory:
+            return redirect(
+                url_for('territories.territory', territory=valid_territory))
+        else:
+            return abort(404)
 
+    DATASETS = TERRITORY_DATASETS[territory.level_name]
     base_dataset_classes = sorted(DATASETS.values(), key=lambda a: a.order)
     base_datasets = [
         base_dataset_class(territory)
         for base_dataset_class in base_dataset_classes
     ]
+    territories = [territory]
+
+    # Deal with territories with ancestors (new/old French regions
+    # for instance).
+    if territory.ancestors:
```

Comment:
Ancestors might be at least an empty list so I think you can drop the `if` and start looping safely